lib/src/custom_watcher_factory.dart

```diff
@@ -1,35 +1,48 @@
-import '../watcher.dart';
+// Copyright (c) 2020, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
 
-/// Defines a way to create a custom watcher instead of the default ones.
-///
-/// This will be used when a [DirectoryWatcher] or [FileWatcher] would be
-/// created and will take precedence over the default ones.
-abstract class CustomWatcherFactory {
-  /// Uniquely identify this watcher.
-  String get id;
+import '../watcher.dart';
 
-  /// Tries to create a [DirectoryWatcher] for the provided path.
-  ///
-  /// Returns `null` if the path is not supported by this factory.
-  DirectoryWatcher createDirectoryWatcher(String path, {Duration pollingDelay});
+/// A factory to produce custom watchers for specific file paths.
+class _CustomWatcherFactory {
+  final String id;
+  final DirectoryWatcher Function(String path, {Duration pollingDelay})
+      createDirectoryWatcher;
+  final FileWatcher Function(String path, {Duration pollingDelay})
+      createFileWatcher;
 
-  /// Tries to create a [FileWatcher] for the provided path.
-  ///
-  /// Returns `null` if the path is not supported by this factory.
-  FileWatcher createFileWatcher(String path, {Duration pollingDelay});
+  _CustomWatcherFactory(
+      this.id, this.createDirectoryWatcher, this.createFileWatcher);
 }
 
 /// Registers a custom watcher.
 ///
-/// It's only allowed to register a watcher factory once per [id] and at most
-/// one factory should apply to any given file (creating a [Watcher] will fail
-/// otherwise).
-void registerCustomWatcherFactory(CustomWatcherFactory customFactory) {
-  if (_customWatcherFactories.containsKey(customFactory.id)) {
-    throw ArgumentError('A custom watcher with id `${customFactory.id}` '
+/// Each custom watcher must have a unique [id] and the same watcher may not be
```

Comment:
Would it be better if we created an ID and returned the value as to prevent registering watchers multiple times?

Reply:
I think it makes the most sense for the definition of the watcher to own the ID. It can be something meaningful like `'citcwatcher'`.